bootloader/assets/grub.go

```diff
@@ -0,0 +1,150 @@
+
+    # use try-kernel.efi
+    set kernel=try-kernel.efi
+elif [ "$kernel_status" = "trying" ]; then
+    # nothing cleared the "trying snap" so the boot failed
+    # we clear the mode and boot normally
+    set kernel_status=""
+    save_env kernel_status
+elif [ -n "$kernel_status" ]; then
+    # ERROR invalid kernel_status state, reset to empty
+    echo "invalid kernel_status!!!"
+    echo "resetting to empty"
+    set kernel_status=""
+    save_env kernel_status
+fi
+
+if [ -e $prefix/$kernel ]; then
+menuentry "Run Ubuntu Core 20" {
+    # use $prefix because the symlink manipulation at runtime for kernel snap
+    # upgrades, etc. should only need the /boot/grub/ directory, not the
+    # /EFI/ubuntu/ directory
+    chainloader $prefix/$kernel snapd_recovery_mode=run $cmdline
+}
+else
+    # nothing to boot :-/
+    echo "missing kernel at $prefix/$kernel!"
+fi
+`
+
+const grubRecoveryScript = `# X-Snapd-boot-script-edition: 1
+
+set default=0
+set timeout=3
+set timeout_style=hidden
+
+if [ -e /EFI/ubuntu/grubenv ]; then
+   load_env --file /EFI/ubuntu/grubenv snapd_recovery_mode snapd_recovery_system
+fi
+
+# standard cmdline params
+set cmdline="console=ttyS0 console=tty1 panic=-1"
+
+# if no default boot mode set, pick one
+if [ -z "$snapd_recovery_mode" ]; then
+    set snapd_recovery_mode=install
+fi
+
+if [ "$snapd_recovery_mode" = "run" ]; then
+    default="run"
+elif [ -n "$snapd_recovery_system" ]; then
+    default=$snapd_recovery_mode-$snapd_recovery_system
+fi
+
+search --no-floppy --set=boot_fs --label ubuntu-boot
+
+if [ -n "$boot_fs" ]; then
+    menuentry "Continue to run mode" --hotkey=n --id=run {
+        chainloader ($boot_fs)/EFI/boot/grubx64.efi
+    }
+fi
+
+# globbing in grub does not sort
+for label in /systems/*; do
+    regexp --set 1:label "/([0-9]*)\$" "$label"
+    if [ -z "$label" ]; then
+        continue
+    fi
+    # yes, you need to backslash that less-than
+    if [ -z "$best" -o "$label" \< "$best" ]; then
+        set best="$label"
+    fi
+    # if grubenv did not pick mode-system, use best one
+    if [ -z "$snapd_recovery_system" ]; then
+        default=$snapd_recovery_mode-$best
+    fi
+    set snapd_recovery_kernel=
+    load_env --file /systems/$label/grubenv snapd_recovery_kernel
+
+    # We could "source /systems/$snapd_recovery_system/grub.cfg" here as well
+    menuentry "Recover using $label" --hotkey=r --id=recover-$label $snapd_recovery_kernel recover $label {
+        loopback loop $2
+        chainloader (loop)/kernel.efi snapd_recovery_mode=$3 snapd_recovery_system=$4 $cmdline
+    }
+    menuentry "Install using $label" --hotkey=i --id=install-$label $snapd_recovery_kernel install $label {
+        loopback loop $2
+        chainloader (loop)/kernel.efi snapd_recovery_mode=$3 snapd_recovery_system=$4 $cmdline
+    }
+done
+
+menuentry 'System setup' --hotkey=f 'uefi-firmware' {
+    fwsetup
+}
+`
+
+func init() {
+	registerAsset("grub.conf", []byte(grubBootScript))
```

Comment:
these should use the target names I think  *.cfg, and then we need to deal with the fact that the current gadget used conf names without symlinks